### PR TITLE
Improve message for discarded pure non-Unit values

### LINKED ISSUE #18408: confusing suggestion when non-Unit expression is adapted to Unit (expression discarded in statement position)
## Compiler version

3.3.0

## Minimized code

```Scala
def fa(f: String ?=> Unit): Unit = f(using "hello")
fa(42) // compiles with warning but does nothing
```

## Output

```scala
[warn] -- [E129] Potential Issue Warning: /Test.scala:13:11 
[warn] 13 |        fa(42) // compiles but does nothing
[warn]    |           ^^
[warn]    |A pure expression does nothing in statement position; you may be omitting necessary parentheses
[warn]    |
[warn]    | longer explanation available when compiling with `-explain`
```

## Expectation

Compilation error. I expect this because `42` seems to not match the type `String ?=> Unit`

### LINKED ISSUE #18722: Warning when assigning null to Unit
## Compiler version

3.3.0

## Minimized example

```Scala
scala> val a: Unit = null
```

## Output Error/Warning message

```scala
1 warning found
-- [E129] Potential Issue Warning: ---------------------------------------------
1 |val a: Unit = null
  |              ^^^^
  |A pure expression does nothing in statement position; you may be omitting necessary parentheses
  |
  | longer explanation available when compiling with `-explain`
val a: Unit = ()
```

## Why this Error/Warning was not helpful

The message was unhelpful because...

...it's hard to understand why it is relevant in this scenario. I initially thought that it's perhaps not allowed to assign `null` to a value of `Unit` type, but since it's just a warning and not an error, then I guess it is. So I don't really understand at all why this warning appears here.

## Suggested improvement

It could be made more helpful by...

...not giving a warning at all, or at least changing the explanation if it's actually a legitimate warning.

### PR DESCRIPTION
**Pull request by nicolasstucki**
Fixes #18408
Fixes #18722

### CODE REVIEW

**Review by bishabosha — COMMENTED** [review]
does this also work for https://github.com/lampepfl/dotty/issues/18408?

**Comment by nicolasstucki** [comment]
> does this also work for https://github.com/lampepfl/dotty/issues/18408?

Yes, it also fixes that issue. I will add it to the tests.

**Review by bishabosha — APPROVED** [review]

**Comment by bishabosha** [comment]
does this still need to be a draft?

**Comment by nicolasstucki** [comment]
> does this still need to be a draft?

I added a second commit with some extra cleanups. Now it is ready for review.

**Review by bishabosha — APPROVED** [review]

File: tests/neg-macros/annot-suspend-cycle.check [hunk 1]
```diff
@@ -1,7 +1,7 @@
 -- [E129] Potential Issue Warning: tests/neg-macros/annot-suspend-cycle/Macro.scala:7:4 --------------------------------
 7 |    new Foo
   |    ^^^^^^^
-  |    A pure expression does nothing in statement position; you may be omitting necessary parentheses
+  |    A pure expression does nothing in statement position
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply tests/neg-macros/annot-suspend-cycle.check:4]: It is correct. Class `Foo` has a pure constructor.

**Pull request merged**